### ISSUE TITLE
Did anything change with respect to generating booleans between 3.51 and 4.0?

### ISSUE THREAD

**Issue by tlycken** [AutoFixture/AutoFixture#973]
We're using AutoFixture quite extensively in our project (thank you for building an invaluable tool!) and recently a bunch of tests started failing. Most of them we've been able to some way or other trace back to AutoFixture generating a `bool` value, which if `true` will execute a code path that blows the test, while `false` will not (sometimes the failing path is executed only if all of a collection of `bool`s are `true`, etc). Many of these tests have been around for months, and started failing last week; they now fail most of the time, but not all of them consistently. We haven't been able to pinpoint under what conditions they pass.

We've also recently upgraded AutoFixture from 3.51.0 to 4.0.0.

Maybe (probably) I'm grasping at straws here, but did something change between those two AutoFixture versions with respect to how boolean values are generated?

I've seen some [earlier][1] [discussions][2] on the topic, but I haven't found anything recent and I don't see anything suspicious in the release notes.

[1]: https://github.com/AutoFixture/AutoFixture/issues/83
[2]: https://github.com/AutoFixture/AutoFixture/issues/322

**Comment by zvirja** [comment]
As far as I know, nothing was changed in regards to the boolean value generation.

The current strategy is to provide the alternating values each time the `bool` value is requested (see implementation [here](https://github.com/AutoFixture/AutoFixture/blob/8bdf92592425aedacf587c127b1fcb70c10d108f/Src/AutoFixture/BooleanSwitch.cs#L35)). Given that switcher is global to the fixture, it's hard or even impossible to predict the generated bool values and guarantee them further as your code evolve.

Consider the following samples:
```c#
public class OtherType
{
    public bool Value { get; set; } // Comment out this line
}

public class TypeWithBooleanProperties  
{
    public OtherType Child { get; set; }
    public bool First { get; set; }
    public bool Second { get; set; }
    public bool Third { get; set; }
}

[Fact]
public void TestBooleanGeneration()
{
    var fixture = new Fixture();
    var result = fixture.Create<TypeWithBooleanProperties>();
}
```

If you check properties of the `result` variable, you'll get the following results (tested with v4.0.1):
```
First: False
Second: True
Third: False
```

However, if you comment out the highlighted line, the result will become opposite:
```
First: True
Second: False
Third: True
```

Therefore, as long as your code changes, the actual values might change as well and you cannot rely on the particular values in your tests. Given that we always have the deeply nested nodes in our object graph, it's almost impossible to track/control the actual values.

Therefore, to make your tests stable, I'd suggest you two ways:
1. In setup of each test always set the value of the boolean property/field to the required value.
2. Customize fixture to always generate static `bool` value (i.e. always `true` or always `false`), but I wouldn't recommend that.

---
Hope this answer helps you somehow. Let me know if you have other questions (or more specific ones).

**Comment by tlycken** [comment]
Thanks for the write-up!

In our tests, we use a new fixture instance for each test, so if I understand you correctly new tests _shouldn't_ interfere with old ones, because they use different fixtures. However, I can reliably make at least some of the tests pass by hard-coding the boolean values to `false`, so there seems to be _something_ that has changed between now and a while ago; it appears that AutoFixture used to give us 100% `false` for those, and now doesn't. We've updated some of our fixture setup as well, so there might be something hidden underneath those layers that we haven't been able to pin-point yet..

I did figure this was a long shot, but it was worth checking if just to be able to rule out one possible source. Thanks again!
